### ISSUE TITLE
Codependency between Constant propagation and ShapeInference

### ISSUE THREAD

**Issue by philass** [onnx/onnx-mlir#1405]
I'm running into the situation where

I get the following error

```
`return emitError("Unable to read scales tensor value");`
```

The problem is that the `scaled` input I'm using to resize is not a Constant but rather a cast of a Constant. So I then try changing the order of my passes such that shape inference runs after constant propagation. That solves the issue for this error, but creates errors in other places.

Some of the Constant propagation assumes that shape inference has already been ran, and segfaults otherwise. Take Transpose for example

```cpp
def CreateTransposeOfConst :
   NativeCodeCall<"ConstPropTranspose($_builder, $0, $1)">;
```

```
def TransposeofConst :  Pat<
    // From TransposeOp(c, p)
    (ONNXTransposeOp:$resOp (ONNXConstantOp:$input $_, $_, $_, $_, $_, $_, $_, $_), $_),
    // To c' where c' is transposed attribute
    (CreateTransposeOfConst $resOp, $input),
    [(IsFromDenseONNXConstantOp:$input)]>;
```

```cpp
ONNXConstantOp ConstPropTranspose(
    PatternRewriter &builder, Value resOperand, Value constValue) { // Value resOperand, Attribute attr, ArrayAttr permAttr
  // Read dense attribute, the constant tensor we are transforming.
  ArrayAttr permAttr = llvm::cast<ONNXTransposeOp>(resOperand.getDefiningOp()).permAttr();
  Attribute attr = llvm::cast<ONNXConstantOp>(constValue.getDefiningOp()).valueAttr();
  DenseElementsAttr denseAttr =
      attr.dyn_cast_or_null<mlir::DenseElementsAttr>();
  assert(denseAttr && "expected dense attribute");
  ShapedType resType = resOperand.getType().cast<RankedTensorType>();  
  auto rank = denseAttr.getType().getShape().size(); // THIS IS THE LINE THAT SegFaults
  ...
```
Here Transpose uses the assumption thats its resultType has already been infered by casting it to a RankedTensorType.

---
So in summary, I'm left with one of two options
- Fail at shape inference because I need to use Constant Propogation
- Fail at Constant Propogation because I need to use Shape Inference

Do you guys have any suggestions for how I should work around this problem? Or a place in the codebase where I could add a change to help address this issue

**Comment by philass** [comment]
For some ops where `inferShapes` can't infer the shape, they just return with `success()` without setting a shape and others emit errors. 

So a possible workaround for me, would be to just return with success(), instead of with the `emitError`.

 I could then run my shape inference pass, then my constant propogation pass and then my shape inference pass one more time.
 
How would you guys feel about that?

**Comment by philass** [comment]
Ignore me, this actually seems to be whats happening in the codebase,

I think my branch has diverged at some point
